### PR TITLE
[Hub Generated] Review request for Microsoft.Insights to add version stable/2018-03-01

### PR DESCRIPTION
**Pull request by adibiton**
If you are a MSFT employee you can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/adibiton/azure-rest-api-specs/dev-monitor-Microsoft.Insights-2018-03-01...master/).

### Contribution checklist:
- [ ] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

File: specification/monitor/resource-manager/Microsoft.Insights/stable/2018-03-01/metricAlert_API.json
Comment (anthony-c-martin) [inline specification/monitor/resource-manager/Microsoft.Insights/stable/2018-03-01/metricAlert_API.json]: What is this - is this an ARM resource?

**Review by anthony-c-martin — APPROVED** [review]

**Comment by ChenTanyi** [comment]
@adibiton Could you respond for the comment and fix the CI error?

File: specification/monitor/resource-manager/Microsoft.Insights/stable/2018-03-01/metricAlert_API.json
Comment (adibiton) [reply specification/monitor/resource-manager/Microsoft.Insights/stable/2018-03-01/metricAlert_API.json]: Yes, it's an ARM resource (microsoft.insights/webtests)

**Comment by adibiton** [comment]
I deployed to this PR since i don't have rights for:
https://github.com/Azure/azure-rest-api-specs/pull/8033

@ArcturusZhang @NullMDR  Can you please approve it?

**Comment by adibiton** [comment]
+@raych1

**Comment by raych1** [comment]
Hi @ArcturusZhang, can you help review the change again?

**Comment by raych1** [comment]
> +@raych1

I confirmed, the model validation failures is a known issue and can be ignored for this PR.

**Comment by ArcturusZhang** [comment]
Hi @adibiton the SDK automation is still reporting problems.
I have a local test to align those two enum definition completely, the generation can finish without errors. I.e I made those two enums have same values and order, as well as same value of modelAsString property (`false`), please have a try.

**Comment by ArcturusZhang** [comment]
And based on the error context, the exception is thrown from `Autorest.Model` namespace which should belong to `autorest.modeler`, I believe this problem is now raised but not in the past because modeler has some kind of updates and becomes more strict on this.

**Comment by adibiton** [comment]
@ArcturusZhang We have no problem changing the enum, however isn't it a breaking change for our swagger?

**Comment by ArcturusZhang** [comment]
> @ArcturusZhang We have no problem changing the enum, however isn't it a breaking change for metricAlert_API swagger? anyway, i pushed the change to metricAlert_API (aligned the enum) - however the errors still remains...

Well, I am not sure whether changing the `modelAsString` value counts as a breaking change, maybe this should be decided by the ARM Review team?
And now only the dotnet SDK generation fails, which is a great progress!

**Comment by ArcturusZhang** [comment]
@NullMDR could you please have a look at the error for dotnet generation? The autorest has succeeded, but the CI fails for some reasons.
@raych1 now ModelValidation reports some problems, please have a look.

**Comment by NullMDR** [comment]
For net generation failure @isra-fel . Seems like it's related to test project?

**Comment by raych1** [comment]
@ArcturusZhang , model validation failures is a known issue and can be ignored for this PR.  @ChenTanyi, I would suggest `modelasstring` change needs ARM re-signoff.

**Comment by isra-fel** [comment]
> For net generation failure @isra-fel . Seems like it's related to test project?

Yes

**Review by ChenTanyi — APPROVED** [review]

**Comment by adibiton** [comment]
@gucalder can you please review?

**Comment by ChenTanyi** [comment]
@anthony-c-martin there is a `modelasstring` change after ARM-signoff, please review it again. Thanks!
@raych1 How to deal with this `ModelValidation` error? Since it is a required CI.

**Comment by raych1** [comment]
> @anthony-c-martin there is a `modelasstring` change after ARM-signoff, please review it again. Thanks!
> @raych1 How to deal with this `ModelValidation` error? Since it is a required CI.

@ChenTanyi , we could force merge this PR as long as there's only the confirmed model validation failures after ARM review again on the change and sign off. Can you remove the ARM signoff tag and ask ARM to review on the last change submitted?

**Comment by ChenTanyi** [comment]
@raych1 Yeah, I have removed the ARM signoff tag and added wait for ARM review tag. Once the arm approves, please force merge it.

**Comment by adibiton** [comment]
@anthony-c-martin Can you review it?

File: specification/monitor/resource-manager/Microsoft.Insights/stable/2018-03-01/metricAlert_API.json
Comment (anthony-c-martin) [reply specification/monitor/resource-manager/Microsoft.Insights/stable/2018-03-01/metricAlert_API.json]: Thanks - would you mind updating the description slightly to reflect this then - "The web test resource Id"?

File: specification/monitor/resource-manager/Microsoft.Insights/stable/2018-03-01/metricAlert_API.json
Comment (anthony-c-martin) [inline specification/monitor/resource-manager/Microsoft.Insights/stable/2018-03-01/metricAlert_API.json]: Why this change?

File: specification/monitor/resource-manager/Microsoft.Insights/stable/2018-03-01/metricAlert_API.json
Comment (anthony-c-martin) [inline specification/monitor/resource-manager/Microsoft.Insights/stable/2018-03-01/metricAlert_API.json]: Adding a new enum field (even "None") is regarded as a breaking change, which requires a new API version. Is this change necessary?

Comment (adibiton) [reply specification/monitor/resource-manager/Microsoft.Insights/stable/2018-03-01/metricAlert_API.json]: @anthony-c-martin I did it in order to be align with:
https://github.com/Azure/azure-rest-api-specs/blob/cf33826d06605cde127d2241e54dd6df55e9145f/specification/monitor/resource-manager/Microsoft.Insights/stable/2018-01-01/metricDefinitions_API.json#L137

I don't mine returning it to the previous condition

File: specification/monitor/resource-manager/Microsoft.Insights/stable/2018-03-01/metricAlert_API.json
Comment (adibiton) [reply specification/monitor/resource-manager/Microsoft.Insights/stable/2018-03-01/metricAlert_API.json]: @anthony-c-martin I did it in order to be align with:

https://github.com/Azure/azure-rest-api-specs/blob/cf33826d06605cde127d2241e54dd6df55e9145f/specification/monitor/resource-manager/Microsoft.Insights/stable/2018-01-01/metricDefinitions_API.json#L137

**Comment by adibiton** [comment]
Hi @gucalder can you please review?

**Comment by ChenTanyi** [comment]
@adibiton you need to change the API based on the ARM review, at least the `"The web test Id."`
Meanwhile, please don't include not necessary breaking change in API change. I think your reply doesn't figure out why your change is necessary.

File: specification/monitor/resource-manager/Microsoft.Insights/stable/2018-03-01/metricAlert_API.json
Comment (adibiton) [reply specification/monitor/resource-manager/Microsoft.Insights/stable/2018-03-01/metricAlert_API.json]: Changed to "The Application Insights web test Id."

**Comment by adibiton** [comment]
@anthony-c-martin we reverted the changes and changed the description - can you please review?

**Comment by ChenTanyi** [comment]
@anthony-c-martin Could you review again after the change as you requested.

**Review by anthony-c-martin — APPROVED** [review]

**Comment by ChenTanyi** [comment]
@raych1 Please force merge as you said. Thanks!

**Comment by raych1** [comment]
@akning-ms , can you please merge this PR since the model validation failures is a known issue?

**Comment by ChenTanyi** [comment]
@raych1 I've assigned to you since I have no permission to force merge. Thanks!

**Comment by ChenTanyi** [comment]
@ArcturusZhang @NullMDR It seems resulting SDK generation error after reverted. I found java didn't generate lite sdk for monitor. What about go and other SDK? How do you deal with the old error?(I think the PR didn't change anything related to the reporting error)

**Comment by ArcturusZhang** [comment]
Those changes on `AggregationType` is made to resolve the autorest failure. If those changes are reverted, the SDK cannot be generated properly.
@anthony-c-martin how could we gracefully resolve this enum item conflict without introducing breaking changes?

**Comment by ChenTanyi** [comment]
@ArcturusZhang One question is that how the old sdk be generated? Since this PR doesn't change for `AggregationType`. Is there another PR change the `AggregationType` after the previous generation?

**Comment by ArcturusZhang** [comment]
> @ArcturusZhang One question is that how the old sdk be generated? Since this PR doesn't change for `AggregationType`. Is there another PR change the `AggregationType` after the previous generation?

One suspect is something changed inside autorest

**Comment by ChenTanyi** [comment]
@ArcturusZhang It seems hard to deal with. Can we just merge it and deal with sdk generation with another PR if needed, since the generation error is nothing to do with this API change.

Otherwise, @anthony-c-martin can we introduce the breaking change you commented? It is used to fix the SDK generation error.

**Comment by raych1** [comment]
Linked the model validation [issue](https://github.com/Azure/oav/issues/468).

**Comment by ChenTanyi** [comment]
@raych1 is it still not fixed yet? The ci is still error.

**Comment by raych1** [comment]
@ChenTanyi , that is expected. I updated the modlevalidation version for PPE pipeline only. Pls check that result which is success now.

**Comment by ChenTanyi** [comment]
@raych1 Then it still need to force merge. Why do you fix a not required ci rather than all ci?

@ArcturusZhang I found that add a enum "None" would fix the sdk generation, (not change the model), it is not a breaking change, could you confirm at Go SDK?

**Comment by ArcturusZhang** [comment]
> @raych1 Then it still need to force merge. Why do you fix a not required ci rather than all ci?
> 
> @ArcturusZhang I found that add a enum "None" would fix the sdk generation, (not change the model), it is not a breaking change, could you confirm at Go SDK?

ARM says it is a breaking change, and has been reverted in commit `b4cb31dc78fc25b951d3ebb68491cfc8373ade99`.
Every SDK except python is complaining about this, which is not caused by go generator itself

**Comment by raych1** [comment]
> @raych1 I've assigned to you since I have no permission to force merge. Thanks!

@chentanyi, please confirm the SDK failures are expected and would not break SDK generation. Then we could ask Arthur to force merge this PR.

> @raych1 Then it still need to force merge. Why do you fix a not required ci rather than all ci?
> 
@ChenTanyi , it still needs the force merge at this point. It is the normal release flow for the new version of tool with breaking change. We will monitor the fix thru pre-production pipeline for a few days.

**Comment by ChenTanyi** [comment]
@ArcturusZhang Got it. Actually the only change with value is not a breaking change. This is what the model as string do. You could see https://github.com/stankovski/AutoRest/blob/master/Documentation/swagger-extensions.md#x-ms-enum. The breaking change is that the commit not only change the model as string to false, but also add a value to enum. Then I would ask @adibiton if he can add only "None" for it and not change the model.

**Comment by ChenTanyi** [comment]
@adibiton Sorry about it. Could you add the "None" again. (Don't change the modelAsString to false) I think it is not a breaking change and @anthony-c-martin can you confirm about it?

**Comment by adibiton** [comment]
@ChenTanyi Added the "None" value to timeAggregation" enum

**Comment by ArcturusZhang** [comment]
Well, I have done all of these tests before, just adding None back will not work. The `modelAsString` flag must also align.

**Comment by ChenTanyi** [comment]
@ArcturusZhang It seems no other solutions about it. @anthony-c-martin The breaking change is needed for SDK generation. Is it OK to making this breaking change https://github.com/Azure/azure-rest-api-specs/pull/7920#discussion_r371377937? Otherwise, is there any solution for resolving it without introducing breaking change?

**Comment by ChenTanyi** [comment]
@adibiton I found another solution about it. Is it possible to remove [this line](https://github.com/Azure/azure-rest-api-specs/pull/7920/files#diff-608869ef7f7185dbc5615c6e3c5d6bbaR830)? The type string is the root cause of sdk generation failure. And the "None" value could be removed if it is not necessary.

**Comment by adibiton** [comment]
> @adibiton I found another solution about it. Is it possible to remove [this line](https://github.com/Azure/azure-rest-api-specs/pull/7920/files#diff-608869ef7f7185dbc5615c6e3c5d6bbaR830)? The type string is the root cause of sdk generation failure. And the "None" value could be removed if it is not necessary.

Pushed a commit - @ChenTanyi Can you please review?

**Comment by ChenTanyi** [comment]
@raych1 The sdk automation issue is resolved. Please force merge it as we said before. I'll assign it to you. Thanks!

**Comment by raych1** [comment]
> @adibiton I found another solution about it. Is it possible to remove this line? The type string is the root cause of sdk generation failure. And the "None" value could be removed if it is not necessary.
> 
> Pushed a commit - @ChenTanyi Can you please review?

@ChenTanyi , enum type and value change is breaking change and need ARM signoff. Is it align to service code?

**Comment by ChenTanyi** [comment]
@raych1 It just revert some changes from this PR, I don't think it need to re-signoff.

**Comment by ChenTanyi** [comment]
> > @adibiton I found another solution about it. Is it possible to remove this line? The type string is the root cause of sdk generation failure. And the "None" value could be removed if it is not necessary.
> > Pushed a commit - @ChenTanyi Can you please review?
> 
> @ChenTanyi , enum type and value change is breaking change and need ARM signoff. Is it align to service code?

I don't know about the service code, you can ask @adibiton

**Pull request merged**
